partner_multi_relation/models/res_partner_relation_all.py

```diff
@@ -1,18 +1,100 @@
 # -*- coding: utf-8 -*-
-# © 2014-2016 Therp BV <http://therp.nl>
+# Copyright 2014-2017 Therp BV <http://therp.nl>
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+# pylint: disable=method-required-super
 """Abstract model to show each relation from two sides."""
+import collections
+import logging
+
 from psycopg2.extensions import AsIs
 
 from openerp import _, api, fields, models
+from openerp.exceptions import ValidationError
 from openerp.tools import drop_view_if_exists
 
 
-PADDING = 10
-_RECORD_TYPES = [
-    ('a', 'Left partner to right partner'),
-    ('b', 'Right partner to left partner'),
-]
+_logger = logging.getLogger(__name__)
+
+
+_last_key_offset = -1
+_specification_register = collections.OrderedDict()
+
+
+def register_select_specification(base_name, is_inverse, select_sql):
+    """Register SELECT clause for rows to be included in view.
+
+    Each SELECT clause must contain a first column in the form:
+        '<base_keyfield> * %%(padding)s + %(key_offset),'
+    The %%(padding)s will be used as a parameter for the
+    cursor.execute function, the %(key_offset) will be replaced
+    by dictionary replacement.
+    The columns for each SELECT clause must be ordered as follows:
+    id - specified as defined above:
+    res_model: model._name of the underlying table,
+    res_id: base key in the underlying table,
+    this_partner_id: must refer to a partner
+    other_partner_id: must refer to a related partner
+    type_id: refers to res.partner.relation.type,
+    date_start: start date of relation if relevant or NULL,
+    date_end: end date of relation if relevant or NULL,
+    %(is_inverse)s as is_inverse: used to determine type_selection_id,
+    """
+    global _last_key_offset
```

Comment:
this will fail for multithreaded servers. We need to think of another mechanism, I'd suggest splitting up the sql statement into function similar to https://github.com/OCA/OCB/blob/10.0/addons/account/report/account_invoice_report.py, then overrides can add what they need without destroying multithreading

Reply:
I don't think it can fail as the variable is only used during module loading to gather all the parts of the statement to create the view. I would be amazed if the loading of one module could be interrupted by the loading of anpther module.